src/charm.py

```diff
@@ -0,0 +1,298 @@
+
+        Returns:
+        The content of the current homeserver.yaml file, str.
+        """
+        synapse_config_path = self._SYNAPSE_CONFIG_PATH
+        container = self._container()
+        if container.exists(synapse_config_path):
+            return self._container().pull(synapse_config_path).read()
+        return None
+
+    def _on_config_changed(self, event):
+        """Handle changed configuration.
+
+        Change this example to suit your needs.
+        If you don't need to handle config, you can remove
+        this method.
+
+        Learn more about config at https://juju.is/docs/sdk/config
+        """
+        # Fetch the new config value
+        log_level = self.model.config["log-level"].lower()
+
+        # Do some validation of the configuration option
+        if self._stored.db_password is not None:
+            if self._current_synapse_config() is None:
+                self._run_migrate_synapse()
+            # The config is good, so update the configuration of the workload
+            container = self.unit.get_container(self._CONTAINER_NAME)
+            # Verify that we can connect to the Pebble API
+            # in the workload container
+            if container.can_connect():
+                # Push an updated layer with the new config
+                container.add_layer(
+                        self._CONTAINER_NAME, self._pebble_layer, combine=True)
+                container.replan()
+
+                logger.debug(
+                        "Log level for synapse changed to '%s'", log_level
+                        )
+                self.unit.status = ActiveStatus()
+            else:
+                # We were unable to connect to the Pebble API,
+                # so we defer this event
+                event.defer()
+                self.unit.status = WaitingStatus("waiting for Pebble API")
+        else:
+            # In this case, the config option is bad,
+            # so block the charm and notify the operator.
+            # generate files in /data if not present
+            event.defer()
+            self.unit.status = WaitingStatus("waiting for db")
+
+    @property
+    def _pebble_layer(self):
+        """Return a dictionary representing a Pebble layer."""
+        return {
+            "summary": "matrix synapse layer",
+            "description": "pebble config layer for matrix synapse",
+            "services": {
+                "synapse": {
+                    "override": "replace",
+                    "summary": "matrix synapse",
+                    "command": "/start.py",
+                    "startup": "enabled",
+                    "environment": self._populate_synapse_env_settings(),
+                }
+            },
+            "checks": {
+                "synapse-ready": {
+                    "override": "replace",
+                    "level": "alive",
+                    "http": {
+                        "url": "http://localhost:8008/health",
```

Comment:
Minor (and I don't know if it was already mentioned): Could "8008" be a constant?

Reply:
makes sense, I'll change it